File: _p1_problems/Linked List Cycle II/Solution.java
public class Solution {
    public ListNode detectCycle(ListNode head) {
        if (head == null) return null;
        ListNode s1 = head, s2 = head;
        do {
            s1 = s1.next;
            s2 = s2.next;
            if (s2 != null) s2 = s2.next;
        } while (s2 != null && s1 != s2);
        if (s2 == null) return null;
        s1 = head;
        while (s1 != s2) {
            s1 = s1.next;
            s2 = s2.next;
        }
        return s1;
    }
}

File: _p1_problems/Linked List Cycle II/Solution.rb
class Solution
  def detectCycle(head)
    slow = fast = head
    while (!fast.nil? && !fast.next.nil? && !fast.next.next.nil?)
      if ((slow = slow.next) == (fast = fast.next.next))
        while (head != slow)
          head = head.next
          slow = slow.next
        end
        return head
      end
    end
    nil
  end
end
